benchmark_cusparselt.py

```diff
@@ -36,31 +36,142 @@ def forward(self, x):
+
+    weight = torch.tensor(mask_entries, dtype=dtype, device=device).view(m, k).cuda()
+    return weight, mask_id_sequence
+
+
+def make_mask(mask_id_sequence):
+    lookup = {
+        0: "0100",
+        1: "1000",
+        2: "1100",
+        3: "1001",
+        4: "1101",
+        5: "1110",
+    }
+
+    my_mask = []
+    small_buf  = []
+    for mask_id in mask_id_sequence:
+        small_buf.insert(0,lookup[mask_id])
+
+        if len(small_buf) == 2:
+            string = "".join(small_buf)
+            sign = -1 if string[0] == "1" else 1
+            mask_val = sign * int(string[1:], 2)
+            my_mask.append(mask_val)
 
-    model = Model(m, k).type(dtype).eval().cuda()
+            small_buf = []
+    return my_mask
 
-    input_tensor = torch.randint(
-        10,
+
+# compare different dtypes
+def compare_dtype(m, k, n, batch_size, dtype):
+    print(m, k, n, batch_size, dtype)
+    model = Model(m, k).cuda().half().eval()
+    # create input tensor
+    input_tensor = torch.randint(2,
         (batch_size, n, k),
         device=model.linear.weight.device,
         dtype=dtype,
     )
 
+    print("input_tensor")
+    print(input_tensor)
 
-    latency = benchmark.Timer(
-        stmt="model(input_tensor)",
-        globals={"input_tensor": input_tensor, "model": model}
-    ).blocked_autorange()
+    weight, _ = gen_two_four_sparse_mask(m, k, dtype)
+    bias = torch.zeros(model.linear.bias.data.shape, dtype=dtype).cuda()
+
+    print("weight: ")
+    print(weight)
+
+    # model.linear.weight.data = weight.float()
+    # model.linear.bias.data = bias.float()
+   
+    num_bytes = weight.nelement() * weight.element_size()
+    compressed_size= num_bytes * 10 // 16 
+    # compressed_size = 1536
+    print(f"weight_compressed: {num_bytes} bytes, mask size: {compressed_size} bytes") 
+    weight_compressed = torch.empty((compressed_size // weight.element_size(), ), 
+                                     dtype=dtype, 
+                                     device=device)
+
+    cslt = torch.classes.cusparselt.CusparseLtLinear(weight_compressed,
+                                                     bias)
+    cslt.set_compressed(weight)
+
+    print("weight compressed")
+    print(weight_compressed[:m*k//2].view(m, -1))
+    print("mask")
+    print(weight_compressed[m*k//2:m*k//2+compressed_size].view(m, -1))
+    s_res = cslt.masked_mm(input_tensor.mT).mT
+
+    res = (torch.matmul(weight.half(), input_tensor.mT.half()) + bias.half()).mT
+    res = res.to(dtype)
+    # model_res = model(input_tensor.half())
+    # assert torch.allclose(model_res, res)
+
+    print("dense result:")
+    print(res)
+
+    print("sparse result:")
+    print(s_res)
+
+    sparse_same_dense = torch.allclose(res, s_res)
+    print(f"dense result - sparse result: {sparse_same_dense}")
+    print(res-s_res)
+
+    # assert torch.allclose(
+    #     s_res.float(), res, rtol=1e-3, atol=1e-3
+    # )
+    # devnull = open("/dev/null", "w")
```

Comment:
@pierrestock For removing cuSPARSELt printouts, you can use lines 150-152 and line 164 afterwards. But be careful, since this will suppress anything written to stdout in between those lines, including debug printouts.